### PR TITLE
JP-1136: Compute scaling for WFSS background subtraction using error-weighted mean

### LINKED ISSUE #8989: Background step unit tests take too long to run
_Issue [JP-3806](https://jira.stsci.edu/browse/JP-3806) was created on JIRA by [Ned Molter](https://jira.stsci.edu/secure/ViewProfile.jspa?name=emolter):_

The unit tests for the background step have very long runtimes compared to all the other unit tests.  This appears to be due to the requirement to do multiple CRDS queries per WFSS test, and tens of different filter/pupil combinations being tested.

It seems unnecessary to do full step runs for every filter here.  The different filters/pupils are not processed differently in the code, and the only difference between them appears to be the reference files that get selected.

Furthermore, the reference backgrounds for the WFSS modes (which are retrieved from reference files) are purpose-built to account for imperfections in the real detector, but the synthetic mock data used for these tests are just random noise, so the fidelity of these backgrounds is not meaningfully tested.

### PR DESCRIPTION
**Pull request by emolter**
Resolves [JP-1136](https://jira.stsci.edu/browse/JP-1136)
Resolves [JP-3806](https://jira.stsci.edu/browse/JP-3806)
Closes #8989 

This PR improves the way the reference background is compared with the data for NIRCAM and NIRISS WFSS observations during the background step, and improves the runtime of the background step tests by removing redundant calls from unnecessary parametrization.  The bulleted points below are the high-level changes:

- The behavior prior to this PR was to sigma-clip the data (taking the middle 50%), sigma-clip the background, take the mean of each, and then do ``data - mean_sci / mean_bkg``.  The new behavior, as of this PR, is:

1. compute 
```
factor = sum(data*model_bkg/var) / sum(model_bkg*model_bkg/var)
sub = data - factor*model_bkg
```
where ``var`` is the sum of all the error terms in the data.  This factor is akin to a variance-weighted mean.
2. Check ``sub`` against a variance criterion to see if it's a good fit.
3. Reject outliers in the data using a sigma-clipping algorithm, initially set so the middle 98% of data values are kept.
4. Repeat steps 1-3 until either a max number of iterations is reached, or the variance did not improve (c.f. a percent threshold) since last iteration.

- Three new top-level step parameters are introduced to control the iterative outlier rejection process during background subtraction - see documentation for the parameter definitions.

- During testing of this change, it was noticed that the WFSS unit tests were not set up in a way that allowed assessment of the goodness-of-fit of the background subtraction.  In short, they were using mock data that just contained noise and comparing it with reference backgrounds from CRDS that contained substantial variations due to real detector imperfections.  This PR makes it so that a fake background with well-known variability is generated on-the-fly, then saved to a file for ``subtract_wfss_bkg`` to read.  A positive side-effect of this change is that it decreases the number of CRDS queries run by the unit tests, so they run faster.

- Unit test parametrizations that caused the background step to be called for every pupil/filter/detector combination were removed, because the code takes an identical path irrespective of the value of these parameters.  The only difference was that different background reference files were found in CRDS, but unit testing is not the place to ensure those queries work correctly (and the tests themselves did not meaningfully test this beyond ensuring that the step runs).  This substantially improved the runtime of these tests.  Calls to each GRISM were retained for each mode because the grism changes the dispersion direction.

- This was definitely beyond the scope of the initial tickets, but it seemed logical to separate the WFSS background subtraction into its own file because it doesn't use any of the same functionality as the background subtraction for the other modes, and the iterative outlier rejection added a fair amount of code.  I went ahead and did that refactor but I'm happy to revert it if desired.

## Tasks
- [x] **request a review from someone specific**, to avoid making the maintainers review every PR
- [x] add a build milestone, i.e. `Build 11.3` (use the [latest build](https://github.com/spacetelescope/jwst/milestones) if not sure)
- [x] Does this PR change user-facing code / API? (if not, label with `no-changelog-entry-needed`)
  - [x] write news fragment(s) in `changes/`: `echo "changed something" > changes/<PR#>.<changetype>.rst` (see below for change types) 
  - [x] update or add relevant tests
  - [x] update relevant docstrings and / or `docs/` page
  - [x] [start a regression test](https://github.com/spacetelescope/RegressionTests/actions/workflows/jwst.yml) and include a link to the running job ([click here for instructions](https://github.com/spacetelescope/RegressionTests/blob/main/docs/running_regression_tests.md))
    - [x] Do truth files need to be updated ("okified")?
      - [ ] **after the reviewer has approved these changes**, run `okify_regtests` to update the truth files
- [ ] if a JIRA ticket exists, [make sure it is resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

<details><summary>news fragment change types...</summary>

- ``changes/<PR#>.general.rst``: infrastructure or miscellaneous change
- ``changes/<PR#>.docs.rst``
- ``changes/<PR#>.stpipe.rst``
- ``changes/<PR#>.datamodels.rst``
- ``changes/<PR#>.scripts.rst``
- ``changes/<PR#>.fits_generator.rst``
- ``changes/<PR#>.set_telescope_pointing.rst``
- ``changes/<PR#>.pipeline.rst``

## stage 1
- ``changes/<PR#>.group_scale.rst``
- ``changes/<PR#>.dq_init.rst``
- ``changes/<PR#>.emicorr.rst``
- ``changes/<PR#>.saturation.rst``
- ``changes/<PR#>.ipc.rst``
- ``changes/<PR#>.firstframe.rst``
- ``changes/<PR#>.lastframe.rst``
- ``changes/<PR#>.reset.rst``
- ``changes/<PR#>.superbias.rst``
- ``changes/<PR#>.refpix.rst``
- ``changes/<PR#>.linearity.rst``
- ``changes/<PR#>.rscd.rst``
- ``changes/<PR#>.persistence.rst``
- ``changes/<PR#>.dark_current.rst``
- ``changes/<PR#>.charge_migration.rst``
- ``changes/<PR#>.jump.rst``
- ``changes/<PR#>.clean_flicker_noise.rst``
- ``changes/<PR#>.ramp_fitting.rst``
- ``changes/<PR#>.gain_scale.rst``

## stage 2
- ``changes/<PR#>.assign_wcs.rst``
- ``changes/<PR#>.badpix_selfcal.rst``
- ``changes/<PR#>.msaflagopen.rst``
- ``changes/<PR#>.nsclean.rst``
- ``changes/<PR#>.imprint.rst``
- ``changes/<PR#>.background.rst``
- ``changes/<PR#>.extract_2d.rst``
- ``changes/<PR#>.master_background.rst``
- ``changes/<PR#>.wavecorr.rst``
- ``changes/<PR#>.srctype.rst``
- ``changes/<PR#>.straylight.rst``
- ``changes/<PR#>.wfss_contam.rst``
- ``changes/<PR#>.flatfield.rst``
- ``changes/<PR#>.fringe.rst``
- ``changes/<PR#>.pathloss.rst``
- ``changes/<PR#>.barshadow.rst``
- ``changes/<PR#>.photom.rst``
- ``changes/<PR#>.pixel_replace.rst``
- ``changes/<PR#>.resample_spec.rst``
- ``changes/<PR#>.residual_fringe.rst``
- ``changes/<PR#>.cube_build.rst``
- ``changes/<PR#>.extract_1d.rst``
- ``changes/<PR#>.resample.rst``

## stage 3
- ``changes/<PR#>.assign_mtwcs.rst``
- ``changes/<PR#>.mrs_imatch.rst``
- ``changes/<PR#>.tweakreg.rst``
- ``changes/<PR#>.skymatch.rst``
- ``changes/<PR#>.exp_to_source.rst``
- ``changes/<PR#>.outlier_detection.rst``
- ``changes/<PR#>.tso_photometry.rst``
- ``changes/<PR#>.stack_refs.rst``
- ``changes/<PR#>.align_refs.rst``
- ``changes/<PR#>.klip.rst``
- ``changes/<PR#>.spectral_leak.rst``
- ``changes/<PR#>.source_catalog.rst``
- ``changes/<PR#>.combine_1d.rst``
- ``changes/<PR#>.ami.rst``

## other
- ``changes/<PR#>.wfs_combine.rst``
- ``changes/<PR#>.white_light.rst``
- ``changes/<PR#>.cube_skymatch.rst``
- ``changes/<PR#>.engdb_tools.rst``
- ``changes/<PR#>.guider_cds.rst``
</details>

### CODE REVIEW

**Comment by emolter** [comment]
initial regtests started here: https://github.com/spacetelescope/RegressionTests/actions/runs/12144260342

**Comment by emolter** [comment]
regtests here https://github.com/spacetelescope/RegressionTests/actions/runs/12283134985

File: jwst/background/background_sub_wfss.py [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+
+    def _update_nans(self, sci, bkg, var, mask):
+        sci[mask] = np.nan
+        bkg[mask] = np.nan
+        var[mask] = np.nan
+
+
+    def _compute_rms_residual(self, sci_sub):
+        """
+        Calculate the RMS of the cross-dispersion background, which is found by taking
```
Comment (Rplesha) [inline jwst/background/background_sub_wfss.py:218]: Calculate the background-subtracted RMS along the dispersion axis, which is found by taking the median profile of the image collapsed along the cross-dispersion axis.

File: jwst/background/background_sub_wfss.py [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+                        mask=~bkg_mask)
+
+    # extract the derived factor and apply it to the unmasked, non-outlier-rejected data
+    subtract_this = factor * bkg_ref.data
+    
+    result = input_model.copy()
+    result.data = input_model.data - subtract_this
+    result.dq = np.bitwise_or(input_model.dq, bkg_ref.dq)
+
+    log.info(f"Average of background image subtracted = {np.nanmean(subtract_this):.3e}")
```
Comment (Rplesha) [inline jwst/background/background_sub_wfss.py:95]: Would it be possible to add in a line to the log here with the factor?:
```
log.info(f"Average of scaled background image = {np.nanmean(subtract_this):.5e}")
log.info(f"Scaling factor = {factor:.5e}")
```

**Review by Rplesha — COMMENTED** [review]
Adding in Gaël's review while we wait for his GitHub account to be added to the spacetelescope group.

The code itself appears correct as written, although he has not had time yet to run a test of the code.

File: docs/jwst/background_step/description.rst [hunk 3]
```diff
@@ -87,7 +87,12 @@ For Wide-Field Slitless Spectroscopy expsoures (NIS_WFSS and NRC_WFSS),
 a background reference image is subtracted from the target exposure.
 Before being subtracted, the background reference image is scaled to match the
 signal level of the WFSS image within background (source-free) regions of the
-image. 
+image. The scaling factor is determined based on the variance-weighted mean
```
Comment (gnoir0t) [inline docs/jwst/background_step/description.rst:90]: ... based on the inverse variance weighted mean of the ratio image between the science data and the reference background, i.e., factor = ...
This factor is equivalent to solving for the scaling constant applied to the reference background that gives the maximum likelihood of matching the science data.

**Review by melanieclarke — COMMENTED** [review]
I have not tracked the conversation for this one, so I'm reviewing mostly for surface-level details.  The algorithm looks plausible, and regtest changes for NIRCam and NIRISS WFSS look reasonable.  

Separating out the WFSS background subtraction into a separate file looks like a good idea to me.  I have a few small suggestions below.

Also, test coverage looks pretty good for the changes, but it would be nice to add a test or two for WFSS at the step level.

File: docs/jwst/background_step/arguments.rst [hunk 4]
```diff
@@ -27,3 +27,21 @@ control the sigma clipping, and are passed as arguments to the astropy
...
+  during determination of the reference background scaling factor. Defaults to 5.
+
+``wfss_p_rms``
+  Only applies to Wide Field Slitless Spectroscopy (WFSS) exposures.
+  If the percentage difference in the RMS of the background-subtracted image
+  between iterations is smaller than this value, stop the iterative outlier
+  rejection process.
+  Defaults to 0, i.e., do all iterations up to ``wfss_maxiter``.
+
+``wfss_p``
```
[the remark this answers is not in the document]
Comment (emolter) [reply docs/jwst/background_step/arguments.rst:43]: This was the suggestion of @gnoir0t so soliciting input.  What about ``wfss_outlier_percent``?

[the remark this answers is not in the document]
Comment (melanieclarke) [reply docs/jwst/background_step/arguments.rst:43]: That sounds fine to me.  And `wfss_rms_threshold` instead of `wfss_p_rms` maybe?

**Comment by emolter** [comment]
@melanieclarke I added some top-level tests as you suggested.  I also went ahead and finished the changes relevant to JP-3806, i.e., to remove the looping over the various filter/pupil/detector combinations - those do not cause the code to do anything different besides retrieve different background reference files, so I don't think unit tests are the place for that.  Let me know what you think.

I'll run another set of regression tests after incorporating any more comments.

File: jwst/background/tests/test_background_wfss.py [hunk 5]
```diff
@@ -0,0 +1,362 @@
...
+    nan_frac = np.sum(np.isnan(sci))/sci.size
+    assert np.isclose(nan_frac, INITIAL_NAN_FRACTION, rtol=1e-2)
+
+    # re-compute mask to ignore "real" sources for tests
+    mask = _mask_from_source_cat(result, wavelenrange)
+
+    shared_tests(sci, mask, data.original_data_mean)
+
+
+# test both filters because they have opposite dispersion directions
```
Comment (melanieclarke) [inline jwst/background/tests/test_background_wfss.py:210]: This comment looks irrelevant for this test now

**Review by melanieclarke — COMMENTED** [review]
Thanks for the new tests!  I'm fine with not iterating over the filters for the WFSS unit test - I agree that doesn't add much.

I'd still like to consider changing the step parameter names, and I'd like to hear from the science team that the algorithm as implemented meets their needs, but I think this otherwise looks fine from a software perspective.

**Comment by emolter** [comment]
> I'd still like to consider changing the step parameter names
Me too, I'll just go ahead and do it.  Will plan to go with whatever comes to mind unless you have suggestions

**Comment by Rplesha** [comment]
I believe @gnoir0t is testing the change on the NIRISS side, but let me know otherwise Gaël, and I can calibrate some NIRISS data to look at today.

@Russell-Ryan and/or @NorPirzkal are you working on calibrating data for NIRCam to test the output scientifically? Do you need any help doing so?

File: jwst/regtest/test_associations_standards.py [hunk 6]
```diff
@@ -90,7 +90,7 @@ def __init__(
     MakePars('pool_030_mir_lrs_nods_bkg'),
     MakePars('pool_031_mir_lrs_nonod_bkg'),
     MakePars('pool_032_nircam_wfss'),
-    MakePars('pool_034_wfss_parallel'),
+    MakePars('pool_034_wfss_outlier_percentarallel'),
```
Comment (melanieclarke) [inline jwst/regtest/test_associations_standards.py:93]: @emolter - this change looks like it was probably unintentional

Comment (emolter) [reply jwst/regtest/test_associations_standards.py:93]: thanks for catching, will fix

**Comment by tapastro** [comment]
Regression tests here: https://github.com/spacetelescope/RegressionTests/actions/runs/12358452874

Expected failures across wfss tests due to change in background subtraction, as well as some failures likely due to connection issues.

File: docs/jwst/background_step/arguments.rst [hunk 7]
```diff
@@ -27,3 +27,21 @@ control the sigma clipping, and are passed as arguments to the astropy
...
+  Only applies to Wide Field Slitless Spectroscopy (WFSS) exposures.
+  Sets the maximum number of iterations allowed for iterative outlier rejection
+  during determination of the reference background scaling factor. Defaults to 5.
+
+``--wfss_rms_stop``
+  Only applies to Wide Field Slitless Spectroscopy (WFSS) exposures.
+  If the percentage difference in the RMS of the background-subtracted image
+  between iterations is smaller than this value, stop the iterative outlier
+  rejection process.
+  Defaults to 0, i.e., do all iterations up to ``wfss_maxiter``.
```
Comment (tapastro) [inline docs/jwst/background_step/arguments.rst:41]: The default described here does not match the default value in the step spec - from the discussion on the ticket, it appears as though this wasn't concretely specified. I think it would be reasonable to set to this to 1, but if you think 0 is more intuitive, I would not object. We just need the two locations to match.

Comment (emolter) [reply docs/jwst/background_step/arguments.rst:41]: This is an RMS stopping criterion, so only stopping when delta RMS is zero seems more intuitive to me.  1 could be a valid value, and it seems that INS wants to run up to the maxiter criterion by default.  But I will check where the default documented here does not equal the spec

Comment (gnoir0t) [reply docs/jwst/background_step/arguments.rst:41]: Yes I also noticed this discrepancy, where in the code the default is set to 1. For the NIRISS tests I've performed, a default of 1 or 0 makes no difference as in both cases the code reaches the default maxiter of 5. But to be safe it could be 0 rather than 1.

**Review by tapastro — APPROVED** [review]
